client/src/controllers/ZoneController.ts

```diff
@@ -0,0 +1,44 @@
+import { Controller } from '@hotwired/stimulus';
+
+/**
+ * Enables the controlled element to respond to specific user interactions
+ * by adding or removing CSS classes dynamically.
+ *
+ * @example
+ * ```html
+ * <div
+ *   data-controller="w-zone"
+ *   data-w-zone-active-class="hovered active"
+ *   data-action="dragover->w-zone#activate dragleave->w-zone#deactivate"
+ * >
+ *   Drag files here
+ * </div>
+ * ```
+ * In above example:
+ * - The `hovered` and `active` classes are applied on `dragover`.
+ * - The classes are removed on `dragleave`.
+ */
+export class ZoneController extends Controller<HTMLElement> {
+  static classes = ['active'];
+
+  /**
+   * `activeClasses` is an array of CSS class names to be:
+   * - Applied when `event->w-zone#activate` is triggered.
+   * - Removed when `event->w-zone#deactivate` is triggered.
+   * These class names are defined in the `data-w-zone-active-class` attribute
+   * and should be space-separated.
+   */
+  declare activeClasses: string[];
+
+  connect() {
+    this.dispatch('ready', { cancelable: false });
+  }
+
+  activate() {
+    this.element.classList.add(...this.activeClasses);
+  }
+
+  deactivate() {
+    this.element.classList.remove(...this.activeClasses);
+  }
```

Comment:
```suggestion

  noop() {}

  }
```

I am going to suggest a way to be able to also attach action (event listeners) for the `document`, to do this we need a method to attach to. Let's add one that does nothing `noop`, we do this in the `ActionController` and it's a handy way to build out Stimulus actions.

Once this is in, we can then adjust our actions by adding the following.

` drop@document->w-zone#noop:prevent`
`dragover@document->w-zone#noop:prevent`

Note: We will still need the `drop`/`dragover` for activate/deactivate ones also.

Notice we are using `@document`, this is Stimulus' way of attaching the event listener to the document. See https://stimulus.hotwired.dev/reference/actions#global-events

Doing this means we can have a something like this to both prevent the default in some but just add/remove classes for others.

e.g.

```
    <div class="drop-zone w-mt-8" data-controller="w-zone" data-w-zone-active-class="hovered" data-action="dragend->w-zone#deactivate dragleave->w-zone#deactivate dragover->w-zone#activate dragover@document->w-zone#noop:prevent drop->w-zone#deactivate drop@document->w-zone#noop:prevent">
```

This should be equivalent to both 
```
  // prevents browser default drag/drop
  $(document).on('drop dragover', function (e) {
    e.preventDefault();
  });
```

and the `core.js` code.
```
$(() => {
  /* Dropzones */
  $('.drop-zone')
    .on('dragover', function onDragOver() {
      $(this).addClass('hovered');
    })
    .on('dragleave dragend drop', function onDragLeave() {
      $(this).removeClass('hovered');
    });
});
```

But, it's best you understand this and set it up yourself to see how it works + make sure it does the right thing.

Reply:
The approach works pretty well but what if create dedicated method for this like:
`preventDefault(event: Event) {
  event.preventDefault();
}`

this will be more readable while defining in action:
`drop@document->w-zone#preventDefault`

if you say so, then I will do it. And for consistency, I would change this in the ActionController too.

Reply:
Yep. We don't want to replace the Stimulus action options, we just want a simpler way to use them when we use this controller. 

I'm seeing this ZoneController as something quite similar to ActionController, letting you do 'utility' like things. 

We may find that long term this will change, but the `noop` has been useful in the action controller. 

We could always also use the action controller here but it's meant to be focused on 'button-ish' stuff.